### PR TITLE
Share TLS config with management, amqp 1.0, (web) mqtt and (web) stomp plugins

### PR DESCRIPTION
**Pull request by ferozjilla**
Relates to https://github.com/rabbitmq/cluster-operator/issues/267

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

## Summary Of Changes

Add TLS support for 
- management
- amqp 1.0
- mqtt and
- stomp plugins 

when TLS support is specified in the custom resource.

This is done by
- opening tls ports for each plugin on the RabbitMQ container
- exposing tls ports for each plugin on the service (formerly the client service)
- extending the `rabbitmq.conf` configmap to enable TLS for each plugin

Also included in these changes:
- renaming `management` to `http`

## Additional Context

The list of plugins considered to share the TLS config have been documented [in the issue](https://github.com/rabbitmq/cluster-operator/issues/267#issuecomment-723028006).

## Testing

Added a system test that checks connectivity of the http API over TLS.

### CODE REVIEW

**Comment by ferozjilla** [comment]
TODO on this PR:
- add system tests for tls over mqtt and stomp

File: internal/resource/configmap_test.go
Comment (ChunyiLyu) [inline internal/resource/configmap_test.go:314]: We can have the description a bit clearer that this test is about testing the tls config for `rabbitmq_web_mqtt` and `rabbitmq_web_stomp`. Right now the spec description reads the same as the one from line `239`.

Comment (ablease) [reply internal/resource/configmap_test.go:314]: Ah good spot. I agree we can be more clear here, thanks Chunyi!

File: internal/resource/configmap_test.go
Comment (Zerpet) [inline internal/resource/configmap_test.go:353]: How are we testing Web STOMP and Web MQTT here?

Following [RabbitMQ docs](https://www.rabbitmq.com/web-stomp.html#tls), I was expecting to see the keys `web_stomp.ssl.certfile` and `web_stomp.ssl.keyfile`. Same story for [MQTT](https://www.rabbitmq.com/web-mqtt.html#tls).

File: internal/resource/configmap_test.go
Comment (Zerpet) [inline internal/resource/configmap_test.go:257]: Neat pick: perhaps we can use `generateRabbitmqCluster()` and modify only the bits an pieces we want to test?

https://github.com/rabbitmq/cluster-operator/blob/911ef0e91ce9028599ee6539529af09397613ae9/internal/resource/statefulset_test.go#L1572-L1579

Something in the lines of:

```golang
instance = generateRabbitmqCluster()
instance.spec.ObjectMeta.Name = "rabbit-tls"
instance.spec.TLS.SecretName = "a-secret"
instance.spec.Rabbitmq.AdditionalPlugins = []rabbitmqv1beta1.Plugin{ ... }
```

**Review by Zerpet — CHANGES_REQUESTED** [review]
The test for secure connections over web sockets doesn't look right to me. Let's double check how to configure those.

File: internal/resource/configmap_test.go
Comment (ferozjilla) [reply internal/resource/configmap_test.go:353]: Web mqtt and web stomp are not implemented yet. The decision to add them too was reached at sync-up and is one of the actions (along with removing the tls property) that we need to do.

File: internal/resource/configmap_test.go
Comment (ferozjilla) [reply internal/resource/configmap_test.go:257]: That's a good idea!

**Comment by ChunyiLyu** [comment]
What's the particular reason to rename port 'management' to 'http'? I think @mkuratczyk  has just recently PR the rename from 'http' to 'management' in the statefulSet: https://github.com/rabbitmq/cluster-operator/pull/444

File: internal/resource/configmap_test.go
Comment (ablease) [reply internal/resource/configmap_test.go]: Hi @Zerpet. configmap additions for web_mqtt and web_stomp have been added here https://github.com/rabbitmq/cluster-operator/pull/451/commits/bc28103a2ba49af6a2d62e6facc1cf054203357c. Please take a look and let me know if it looks right.

File: internal/resource/configmap_test.go
Comment (ablease) [reply internal/resource/configmap_test.go:257]: I like this idea too. In the interest of time I will create another issue to refactor the config map tests. Thanks Aitor

**Comment by ablease** [comment]
> #444

Ah well spotted. I will opt to rename ports from http to management. What would be a good name for 'https' ports? 'management-tls' perhaps?

**Comment by ablease** [comment]
> What's the particular reason to rename port 'management' to 'http'? I think @mkuratczyk has just recently PR the rename from 'http' to 'management' in the statefulSet: #444

Ah well spotted. I will opt to rename ports from http to management. What would be a good name for 'https' ports? 'management-tls' perhaps?

**Comment by ChunyiLyu** [comment]
> Ah well spotted. I will opt to rename ports from http to management. What would be a good name for 'https' ports? 'management-tls' perhaps?

Sounds good. I see that we are naming tls port as "web-mqtt-tls" and "web-stomp-tls", so lets keep it in the same format.

File: internal/resource/statefulset.go
Comment (ChunyiLyu) [inline internal/resource/statefulset.go:440]: Lets remove the comment here since we've already created an issue for this

File: internal/resource/service.go
Comment (Zerpet) [inline internal/resource/service.go:187]: The function `MutualTLSEnabled()` already checks if `TLSEnabled()`. Don't we get the same result if we just check for `MutualTLSEnabled()` ?

**Review by Zerpet — APPROVED** [review]

File: internal/resource/statefulset.go
Comment (ablease) [reply internal/resource/statefulset.go]: Makes sense, thanks for pointing that out! https://github.com/rabbitmq/cluster-operator/pull/451/commits/b737ca4a59b2e05ac2ba106efb485dbe6f603fcd

File: internal/resource/service.go
Comment (ablease) [reply internal/resource/service.go]: Ah so it does! Will just use the single check here instead. Thanks for pointing that out! Change made here https://github.com/rabbitmq/cluster-operator/pull/451/commits/b737ca4a59b2e05ac2ba106efb485dbe6f603fcd.

File: internal/resource/configmap.go
Comment (ansd) [inline internal/resource/configmap.go:130]: Should we extract some of these paths into variables?
`/etc/rabbitmq-tls/ca.crt` is used 4 times here.

File: system_tests/utils.go
Comment (ansd) [inline system_tests/utils.go:773]: Can we also add a test for WebMQTT and TLS, i.e. testing `wss`?

**Review by ansd — CHANGES_REQUESTED** [review]

File: internal/resource/service.go
Comment (ChunyiLyu) [reply internal/resource/service.go]: updated. Good catch david!

File: controllers/reconcile_tls.go
Comment (ChunyiLyu) [reply controllers/reconcile_tls.go:56]: The error should already contain the secret name and namespace it tries to get but failed.

File: internal/resource/configmap.go
Comment (ChunyiLyu) [reply internal/resource/configmap.go]: updated. Good suggestion!

File: controllers/reconcile_tls.go
Comment (ansd) [reply controllers/reconcile_tls.go:56]: Where is that happening?
The Event contains secret name and namespace, but the error which is returned doesn't?

Comment (ChunyiLyu) [reply controllers/reconcile_tls.go]: @ansd updated. Could you take a look again?

File: system_tests/utils.go
Comment (ChunyiLyu) [reply system_tests/utils.go:773]: I don't see too much value in us testing every single plugins protocol with TLS. I think the coverage we have currently is enough, and would prefer to leave the system tests as it is at least for now.

File: controllers/reconcile_tls.go
Comment (ansd) [reply controllers/reconcile_tls.go]: looks good thanks

File: system_tests/utils.go
Comment (ansd) [reply system_tests/utils.go:773]: The additional value is that we configured MQTT Web Stomp TLS correctly, i.e. the changes in https://github.com/rabbitmq/cluster-operator/pull/451/files#diff-8963e26fa961496c5fbf37e06487c020529c5cee6f6ff953ed8d63c018d1c5efR115-R128 and that we expose the port. We do have unit tests, but an end-to-end test adds value from my point of view that this feature actually works.

However, as discussed in standup, let's not block merging the PR on this test. @ablease said he will add a test once this is merged.

**Review by ansd — APPROVED** [review]

**Pull request merged**
